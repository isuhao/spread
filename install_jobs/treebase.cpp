#include "treebase.hpp"
#include <parent_job/andjob.hpp>
#include <stdexcept>

using namespace Spread;

TreeBase::TreeBase(TreeOwner &o, IHashFinder &f)
  : owner(o), finder(f)
{}

TreeBase::TreeBase(TreeOwner &o)
  : owner(o), finder(o.finder)
{}

std::string TreeBase::setStatus(const std::string &msg)
{
  log("STATUS: " + msg);
  return setBusy(msg);
}
void TreeBase::fail(const std::string &msg)
{
  log("ERROR: " + msg);
  throw std::runtime_error(msg);
}
void TreeBase::log(const std::string &msg)
{ owner.log(msg); }


void TreeBase::addOutput(const Hash &h, const std::string &where) { assert(0); }
void TreeBase::addInput(const Hash &h) { assert(0); }

void TreeBase::fetchFiles(const HashDir &outputs, HashMap &results)
{
  std::string oldMsg = setStatus("Setting up child jobs");

  std::map<Hash,JobInfoPtr> waitList;
  AndJob *aj = NULL;

  {
    std::map<Hash,TreePtr> unpackers;
    HashDir::const_iterator it;

    TreeOwner::Lock lock = owner.lock();
    for(it = outputs.begin(); it != outputs.end(); it++)
      {
        const Hash &hash = it->first;
        const std::string &outfile = it->second;

        // Abort if we have already looked up this hash
        if(waitList.find(hash) != waitList.end())
          continue;

        // Check if there is a system-wide job already running for
        // this target.
        {
          JobInfoPtr inf = owner.getRunningTarget(hash);
          if(inf)
            {
              waitList[hash] = inf;
              continue;
            }
        }

        HashSource src;
        if(!finder.findHash(hash, src, outfile))
          fail("No source for target " + hash.toString() + " " + outfile);
        assert(src.hash == hash);

        // If the file already exists, deal with it later.
        if(src.type == TST_File || src.type == TST_InPlace)
          continue;

        // Probably use a shared_ptr here. And actually, like
        // before we make a Target first, then set it up to do
        // what we want.
        TreePtr job;
        if(src.type == TST_Download)
          job = owner.downloadTarget(src.value);
        else if(src.type == TST_Archive)
          {
            assert(!src.dirHash.isNull());
            assert(src.value == "");
            assert(src.deps.size() == 1);
            Hash arcHash = src.deps[0];

            // Check if we have already created an archive
            // unpacker for this archive
            if(unpackers.find(arcHash) != unpackers.end())
              {
                // Add this target to the output list of the
                // already existing unpacker.
                unpackers[arcHash]->addOutput(hash, outfile);
                continue;
              }

            job = owner.unpackTarget(src.dirHash);
            unpackers[arcHash] = job;
          }
        else assert(0);

        assert(job);

        // Add this target to the jobs output list
        job->addOutput(hash, outfile);

        // Add input dependencies
        for(int i=0; i<src.deps.size(); i++)
          job->addInput(src.deps[i]);

        // Let the outside world know about our target
        owner.setRunningTarget(hash, job->getInfo());

        // And add the job to our local execution list
        if(!aj) aj = new AndJob;
        aj->add(job);
      }
    // 'lock' goes out of scope here
  }

  // This will run all our jobs, and automatically fail if any of them fail.
  if(aj)
    execJob(aj);

  // Wait for external jobs to finish
  std::map<Hash,JobInfoPtr>::const_iterator wit;
  for(wit = waitList.begin(); wit != waitList.end(); wit++)
    {
      setStatus("Waiting for target " + wit->first.toString());
      wit->second->wait(getInfo());
    }

  // Finally check all the cache values, and set up 'results'
  // based on what we find. Fail if anything is missing.
  HashDir::const_iterator it;
  results.clear();
  for(it = outputs.begin(); it != outputs.end(); it++)
    {
      const Hash &hash = it->first;
      const std::string &outfile = it->second;

      // Look up and check if the requested file has been created
      // somewhere.
      HashSource src;
      if(!finder.findHash(hash, src, outfile) ||
         (src.type != TST_File && src.type != TST_InPlace))
        fail("Failed to create target " + hash.toString() + " " + outfile);
      assert(src.hash == hash);
      assert(src.value != "");

      // Copy the file into place if necessary
      if(src.type == TST_File && outfile != "")
        {
          TreePtr job = owner.copyTarget(src.value);
          job->addOutput(hash, outfile);
          execJob(job);
          src.value = outfile;
        }
      assert(src.value != "");

      // Store the result
      results[hash] = src.value;
    }

  setStatus(oldMsg);
}
