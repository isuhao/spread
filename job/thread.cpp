#include "thread.hpp"

#include <boost/thread.hpp>
#include <assert.h>

using namespace Jobify;

struct ThreadObj
{
  Job *j;
  void operator()()
  {
    j->run();
    assert(j->getInfo()->isFinished());
    delete j;
  }
};

void Thread::sleep(double seconds)
{
  int msecs = (int)(seconds*1000000);
  boost::this_thread::sleep(boost::posix_time::microseconds(msecs));
}

void Thread::run(Job *j, bool async)
{
  ThreadObj to;
  to.j = j;
  if(async) boost::thread trd(to);
  else to();
}
