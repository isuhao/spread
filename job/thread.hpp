#ifndef __JOBIFY_THREAD_HPP
#define __JOBIFY_THREAD_HPP

#include "job.hpp"

namespace Jobify
{
  /* This black-box module swallows any Job object, executes it in a
     separate thread, then deletes the object after it returns. (The
     job must have been allocated using 'new'.)

     You can disable threading (run it locally) by setting
     async=false. This may seem pointless compared to just running
     j->run(), but it does provide a consistent interface for both
     cases, and it will also delete the Job before returning.
   */
  struct Thread
  {
    static void run(Job *j, bool async=true);
    static void sleep(double seconds);
  };
}
#endif
