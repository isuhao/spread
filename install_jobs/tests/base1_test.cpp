#include "treebase.hpp"

#include <iostream>
#include <boost/thread/recursive_mutex.hpp>
#include <job/thread.hpp>

using namespace std;
using namespace Spread;

Hash empty;
Hash nofile("nofile");
Hash hello("hello", 5), world("world", 5);
Hash keiko("keiko", 5), postei("postei", 6);

Hash multiArc("ARCME");
Hash a1("A1"), a2("A2"), a3("A3");

float dlsleep = 0, arcsleep = 0;

struct DummyFind : IHashFinder
{
  TreeBase::HashMap files;
  void reset() { files.clear(); }

  bool findHash(const Hash &hash, HashSource &out,
                const std::string &target="")
  {
    cout << "findHash(hash=" << hash << ", target=" << target << ")\n";
    out.hash = hash;
    out.type = TST_None;
    out.dirHash.clear();
    out.value.clear();
    out.deps.clear();

    if(files[hash] != "")
      {
        out.type = TST_File;
        out.value = files[hash];
        if(out.value == target)
          out.type = TST_InPlace;
        return true;
      }

    if(hash == hello || hash == world || hash == multiArc)
      {
        out.type = TST_File;
        out.value = "cache/" + hash.toString().substr(0,4);
        if(target == out.value)
          out.type = TST_InPlace;
        return true;
      }

    if(hash == keiko)
      {
        out.type = TST_Download;
        out.value = "url://SOME/URL/";
        return true;
      }

    if(hash == postei)
      {
        out.type = TST_Archive;
        out.dirHash = postei;
        out.deps.push_back(keiko);
        return true;
      }

    if(hash == a1 || hash == a2 || hash == a3)
      {
        out.type = TST_Archive;
        out.dirHash = postei;
        out.deps.push_back(multiArc);
        return true;
      }

    return false;
  }

  void brokenURL(const Hash &hash, const std::string &url)
  {
    cout << "brokenURL(" << hash << ", " << url << ")\n";
  }

  void addToCache(const Hash::DirMap &dir)
  {
    cout << "Adding " << dir.size() << " files to cache:\n";
    Hash::DirMap::const_iterator it;
    for(it = dir.begin(); it != dir.end(); it++)
      {
        const std::string &file = it->first;
        const Hash &hash = it->second;
        assert(!hash.isNull());
        assert(file != "");
        cout << "  " << hash << " " << file << endl;

        files[hash] = file;
      }
  }
};

struct DummyTarget : TreeBase
{
  std::string what;
  HashDir outs, ins;
  int type;

  DummyTarget(TreeOwner &o, const std::string &w, int t)
    : TreeBase(o), what(w), type(t)
  {
    cout << "Creating target WHAT=" << what << endl;
  }

  void addOutput(const Hash &h, const std::string &where)
  { outs.insert(HDValue(h,where)); }
  void addInput(const Hash &h)
  { ins.insert(HDValue(h,"")); }

  void doJob()
  {
    assert(finder);
    if(type == 1 && dlsleep != 0)
      {
        Thread::sleep(dlsleep);
        cout << "-- end dlsleep\n";
      }
    if(type == 2 && arcsleep != 0)
      {
        Thread::sleep(arcsleep);
        cout << "-- end arcsleep\n";
      }

    cout << "TARGET: " << what << endl;
    if(ins.size())
      {
        cout << "  Inputs:\n";
        {
          HashDir::const_iterator it;
          for(it = ins.begin(); it != ins.end(); it++)
            {
              const Hash &hash = it->first;
              std::string name = it->second;

              cout << "    " << hash << " " << name << endl;
            }
        }

        HashMap res;
        fetchFiles(ins, res);
        cout << "  Processed inputs:\n";
        {
          HashMap::const_iterator it;
          for(it = res.begin(); it != res.end(); it++)
            cout << "    " << it->first << " " << it->second << endl;
        }
      }
    if(outs.size())
      {
        cout << "  Outputs:\n";
        HashDir::const_iterator it;
        Hash::DirMap dir;
        for(it = outs.begin(); it != outs.end(); it++)
          {
            const Hash &hash = it->first;
            std::string name = it->second;

            cout << "    " << hash << " " << name << endl;
            if(name == "")
              name = owner.getTmpName(hash);
            dir[name] = hash;
          }
        finder->addToCache(dir);
        owner.notifyFiles(dir);
      }
    setDone();
  }
};

struct DummyOwner : TreeOwner
{
  TreePtr target(const std::string &what, int type)
  { return TreePtr(new DummyTarget(*this, what, type)); }

  TreePtr copyTarget(const std::string &from)
  { return target("COPY " + from, 0); }

  TreePtr downloadTarget(const std::string &url)
  { return target("DOWNLOAD " + url, 1); }

  TreePtr unpackTarget(const Hash &dir)
  { return target("UNPACK", 2); }

  TreePtr unpackBlindTarget(const string&, const Hash &)
  { assert(0); }

  boost::recursive_mutex mutex;
  Lock lock() { return Lock(new boost::lock_guard<boost::recursive_mutex>(mutex)); }

  void loadDir(const std::string &file, Hash::DirMap &output,
               const Hash &check = Hash())
  { assert(0); }
  void storeDir(const Hash::DirMap &dir, const Hash &check = Hash())
  { assert(0); }

  void notifyFiles(const Hash::DirMap &dir)
  {
    Lock l = lock();
    cout << "notifyFiles():\n";
    Hash::DirMap::const_iterator it;
    for(it = dir.begin(); it != dir.end(); it++)
      {
        const std::string &file = it->first;
        const Hash &hash = it->second;
        cout << "  " << hash << endl;
        stuff[hash] = JobInfoPtr();
      }
  }

  void log(const std::string &msg) { cout << "LOG: " << msg << endl; }

  std::map<Hash, JobInfoPtr> stuff;

  void reset()
  {
    stuff.clear();
  }

  std::string getTmpName(const Hash &hash)
  { return "tmp_" + hash.toString().substr(0,6); }

  JobInfoPtr getRunningTarget(const Hash &hash)
  {
    Lock l = lock();
    cout << "getRunningTarget(" << hash << ")\n";
    JobInfoPtr ptr = stuff[hash];
    if(ptr) cout << "  FOUND!\n";
    return ptr;
  }
  void setRunningTarget(const Hash &hash, JobInfoPtr ptr)
  {
    Lock l = lock();
    cout << "setRunningTarget(" << hash << ")\n";
    assert(ptr);
    stuff[hash] = ptr;
  }
};

struct MyBase : TreeBase
{
  DummyFind *fnd;
  DummyOwner own;

  MyBase() : TreeBase(own)
  {
    fnd = new DummyFind;
    finder.reset(fnd);

    // Cheat and set busy status even though we are not running.
    setBusy();
  }

  // Circumvent protected function for testing purposes
  void fetch(const HashDir &outputs, HashMap &results)
  { fetchFiles(outputs, results); }

  void doJob() { assert(0); }

  void reset() { fnd->reset(); own.reset(); }
};

TreeBase::HashDir makeList;

void add(const Hash &h, const std::string &file="")
{
  makeList.insert(TreeBase::HDValue(h, file));
}

MyBase base;

void test()
{
  try
    {
      TreeBase::HashMap res;

      cout << "\nRUNNING fetchFiles():\n";
      base.fetch(makeList, res);

      cout << "Results returned:\n";
      TreeBase::HashMap::const_iterator it;
      for(it = res.begin(); it != res.end(); it++)
        cout << it->first << " " << it->second << endl;
    }
  catch(std::exception &e)
    {
      cout << "ERROR: " << e.what() << endl;
    }

  makeList.clear();
  base.reset();
}

int main()
{
  add(empty);
  test();

  add(nofile, "file");
  test();

  add(nofile);
  test();

  add(hello, "out_hello");
  add(hello, "out_hello2");
  add(world, "out_world");
  test();

  add(hello, "cache/LPJN");
  add(hello, "");
  add(world, "cache/SG6k");
  add(world, "");
  test();

  add(keiko, "");
  test();

  add(keiko, "blah");
  test();

  add(keiko, "");
  add(keiko, "blah");
  test();

  add(postei, "postei");
  test();

  dlsleep = 0.1;
  add(keiko, "keiko");
  add(postei, "postei");
  test();

  dlsleep = 0;
  arcsleep = 0.1;
  add(keiko, "keiko");
  add(postei, "postei");
  test();

  arcsleep = 0;
  dlsleep = 0;

  add(a1, "__a1_file");
  add(a2, "__a2_file");
  add(a3, "__a3_file");
  add(a2, "__a2_file2");
  test();

  return 0;
}
