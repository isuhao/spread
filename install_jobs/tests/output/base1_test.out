
RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(00)
findHash(hash=00, target=)
LOG: ERROR: No source for target 00 
ERROR: No source for target 00 

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(nofile)
findHash(hash=nofile, target=file)
LOG: ERROR: No source for target nofile file
ERROR: No source for target nofile file

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(nofile)
findHash(hash=nofile, target=)
LOG: ERROR: No source for target nofile 
ERROR: No source for target nofile 

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=out_hello)
getRunningTarget(LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=out_hello2)
getRunningTarget(SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF)
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=out_world)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=out_hello)
TARGET: COPY cache/LPJN
  Outputs:
    LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF out_hello
Adding 1 files to cache:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF out_hello
notifyFiles():
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=out_hello2)
TARGET: COPY out_hello
  Outputs:
    LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF out_hello2
Adding 1 files to cache:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF out_hello2
notifyFiles():
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=out_world)
TARGET: COPY cache/SG6k
  Outputs:
    SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF out_world
Adding 1 files to cache:
  SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF out_world
notifyFiles():
  SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF
LOG: STATUS: Setting up child jobs
Results returned:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF out_hello2
SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF out_world

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=cache/LPJN)
getRunningTarget(LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=)
getRunningTarget(SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF)
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=cache/SG6k)
getRunningTarget(SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF)
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=cache/LPJN)
findHash(hash=LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF, target=)
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=cache/SG6k)
findHash(hash=SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF, target=)
LOG: STATUS: Setting up child jobs
Results returned:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF cache/LPJN
SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF cache/SG6k

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F tmp_U_l4zp
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
LOG: STATUS: Setting up child jobs
Results returned:
U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F tmp_U_l4zp

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=blah)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=blah)
LOG: STATUS: Setting up child jobs
Results returned:
U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
  FOUND!
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F tmp_U_l4zp
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
LOG: STATUS: Waiting for target U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=blah)
TARGET: COPY tmp_U_l4zp
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
LOG: STATUS: Setting up child jobs
Results returned:
U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F blah

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
setRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
TARGET: UNPACK
  Inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F tmp_U_l4zp
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
LOG: STATUS: 
  Processed inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F tmp_U_l4zp
  Outputs:
    QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
Adding 1 files to cache:
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
notifyFiles():
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
LOG: STATUS: Setting up child jobs
Results returned:
QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
setRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=keiko)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
TARGET: UNPACK
  Inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
  FOUND!
LOG: STATUS: Waiting for target U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
-- end dlsleep
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
LOG: STATUS: 
  Processed inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
  Outputs:
    QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
Adding 1 files to cache:
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
notifyFiles():
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=keiko)
LOG: STATUS: Setting up child jobs
Results returned:
QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
setRunningTarget(QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG)
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=keiko)
setRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
TARGET: DOWNLOAD url://SOME/URL/
  Outputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
Adding 1 files to cache:
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
notifyFiles():
  U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F
-- end arcsleep
TARGET: UNPACK
  Inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F 
LOG: STATUS: Setting up child jobs
getRunningTarget(U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=)
LOG: STATUS: 
  Processed inputs:
    U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko
  Outputs:
    QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
Adding 1 files to cache:
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
notifyFiles():
  QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG
findHash(hash=QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG, target=postei)
findHash(hash=U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F, target=keiko)
LOG: STATUS: Setting up child jobs
Results returned:
QSxQXsnmaVwgjK6xVdZP3H-4eeh37BsQMDs1GMe7_aoG postei
U_l4zpQIWVSXtEhsQjGQxUwvlNRH9p5TeK2QwTd30p4F keiko

RUNNING fetchFiles():
LOG: STATUS: Setting up child jobs
getRunningTarget(A1)
findHash(hash=A1, target=__a1_file)
setRunningTarget(A1)
getRunningTarget(A2)
findHash(hash=A2, target=__a2_file)
getRunningTarget(A2)
findHash(hash=A2, target=__a2_file2)
getRunningTarget(A3)
findHash(hash=A3, target=__a3_file)
TARGET: UNPACK
  Inputs:
    ARCME 
LOG: STATUS: Setting up child jobs
getRunningTarget(ARCME)
findHash(hash=ARCME, target=)
findHash(hash=ARCME, target=)
LOG: STATUS: 
  Processed inputs:
    ARCME cache/ARCM
  Outputs:
    A1 __a1_file
    A2 __a2_file
    A2 __a2_file2
    A3 __a3_file
Adding 4 files to cache:
  A1 __a1_file
  A2 __a2_file
  A2 __a2_file2
  A3 __a3_file
notifyFiles():
  A1
  A2
  A2
  A3
findHash(hash=A1, target=__a1_file)
findHash(hash=A2, target=__a2_file)
TARGET: COPY __a2_file2
  Outputs:
    A2 __a2_file
Adding 1 files to cache:
  A2 __a2_file
notifyFiles():
  A2
findHash(hash=A2, target=__a2_file2)
TARGET: COPY __a2_file
  Outputs:
    A2 __a2_file2
Adding 1 files to cache:
  A2 __a2_file2
notifyFiles():
  A2
findHash(hash=A3, target=__a3_file)
LOG: STATUS: Setting up child jobs
Results returned:
A1 __a1_file
A2 __a2_file2
A3 __a3_file
